design/ceph/admission-controller.md

```diff
@@ -0,0 +1,148 @@
+# Admission Controllers in Rook
+
+# Background
+
+This proposal is to add support for admission controllers in Rook. An admission controller is a piece of code that intercepts requests to the Kubernetes API server prior to persistence of the object, but after the request is authenticated and authorized
+
+ There are two special controllers: MutatingAdmissionWebhook and ValidatingAdmissionWebhook.
+ 
+ Mutating controllers may modify the objects they admit but validation controllers are only allowed to validate requests.
+
+Currently, user can manipulate CRD specs with any values which may result in Rook not functioning as expected. The present validation method in Kubernetes is the OpenAPIValidation schema validation which can be used for basic validations like checking type of data, providing a range for the values etc but anything more complex (checking resource availability, network status, error handling) would not be possible under this scenario.
+ 
+# Admission Controllers Concept
+
+![AC Image](ac_design.png)
+
+## Requirements
+
+1. Webhook server which will validate the requests
+2. TLS certificates for the server,
+3. ValidatingWebhookConfig/MutatingWebhookConfig which will intercept requests and send a HTTPS request to the webhook server.
+4. RBAC Components 
+
+## Implementation 
+
+As shown in the above diagram, the admission control process proceeds in two phases. In the first phase, mutating admission controllers are run. In the second phase, validating admission controllers are run. Note again that some of the controllers are both.
+
+The admission controllers intercept requests based on the values given in the configuration.In this config, We have to provide the details on 
+* What resources should it be looking for ? (Pods, Service)
+* What api version and group does it belong to ? *Example : ApiVersion = (v1, v1beta) Group version = (rook.ceph.io, admissionregistration.k8s.io)*
+* What kind of operation should it intercept ? (Create, Update, Delete)
+* A valid base64 encoded CA bundle.
+* What path do we want to send with HTTPs request (/validate, /mutate)
+
+A server should be in place (*with valid TLS certificates*) to intercept any HTTPs request that comes with the above path value. Once intercepted by a server, an [AdmissionRequest](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L40) object is sent through with the resource specifications. 
+
+Here is the part where the server will perform all our validations on the provided resources values and send back an [AdmissionResponse](https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go#L116) with the field Allowed set accordingly to indicate whether it is accepted or rejected.
+
+If any of the controllers in either phase reject the request, the entire request is rejected immediately and an error is returned to the end-user.
+
+## Certificate Management
+
+### Development and Testing 
+
+![Dev_AC_Image](./ac_dev_design.png)
```

Comment:
This diagram is incomplete,
1. MutatingWebhookConfig is not shown.
2. `/validate` endpoint is called twice. It should be once per request.
3. Terms mentioned in boxes like `service` and `webhook` should be self-explanatory.

Reply:
1. Added MutatingWebhookConfig label on the same as ValidatingWebhookConfig as both of them are doing the same, just sending different paths
2. /validate request would be sent once to our kubernetes service which in turn would pass it to our server. No request would be sent twice. Thanks for this suggestion, will make it more clear.
3. I have labeled service as Kubernetes service endpoint and webhook as HTTPS Webhook Server to make it more clear. Let me know if you have any other suggestions